File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int lMin[] = new int[height.length];
        int rMin[] = new int[height.length];
        int res = 0;
        lMin[0] = height[0];
        for(int i=1;i<height.length;i++){
            lMin[i] = Math.max(height[i], lMin[i-1]);
        }
        rMin[height.length-1] = height[height.length - 1];
        for(int j=height.length-2;j>=0;j--){
            rMin[j] = Math.max(height[j], rMin[j+1]);
        }
        for(int k=0;k<height.length;k++){
            res = res + (Math.min(lMin[k],rMin[k]) - height[k]);
        }
        return res;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // O(n^2)
        
        // int res=0;
        // for(int i=1;i<height.size()-1;i++){
        //     int lmax = height[i]; 
        //     for(int j=0;j<i;j++){
        //         lmax = max(height[j],lmax);
        //     } 
        //     int rmax = height[i];
        //     for(int k=i+1;k<height.size();k++){
        //         rmax = max(height[k],rmax);
        //     }
        //     res = res+(min(lmax,rmax)-height[i]);
        // }
        // return res;
        
        // O(n)
        
        int res = 0;
        int larr[height.size()];
        int rarr[height.size()];
        larr[0] = height[0];
        for(int i=1;i<height.size();i++){
            larr[i] = max(height[i],larr[i-1]);
        }
        rarr[height.size()-1] = height[height.size()-1];
        for(int i=height.size()-2;i>=0;i--){
            rarr[i] = max(height[i],rarr[i+1]);
        }
        for(int i=0;i<height.size();i++){
            res = res + (min(larr[i],rarr[i]) - height[i]);
        }
        return res;
    }
    
};